cli_tools/common/vdisk/file_inspector.go

```diff
@@ -0,0 +1,124 @@
+//  Copyright 2020 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package vdisk
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/gcsfuse"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/files"
+
+	"github.com/cenkalti/backoff/v4"
+)
+
+const bytesPerBG = int64(1024 * 1024 * 1024)
+
+// VirtualDiskFileMetadata contains metadata about a virtual disk file.
+type VirtualDiskFileMetadata struct {
+	// Gigabytes used by the file itself, rounded up to the nearest GB.
+	PhysicalSizeGB int64
+
+	// Gigabytes used by the disk, after inflation. Rounded up to the nearest GB.
+	VirtualSizeGB int64
+
+	// The file format used for encoding the VM disk.
+	FileFormat Format
+}
+
+// VirtualDiskFileInspector returns metadata about virtual disk files.
+type VirtualDiskFileInspector interface {
+	// Inspect returns VirtualDiskFileMetadata for the virtual disk file associated
+	// with a reference. IO operations will be retried until the context is cancelled.
+	Inspect(ctx context.Context, reference string) (VirtualDiskFileMetadata, error)
+}
+
+// NewGCSInspector returns a virtualDiskFileInspector that inspects virtual
+// disk files that are stored in the GCS bucket. The Inspect method expects
+// a GCS URI to the file to be inspected.
+func NewGCSInspector() VirtualDiskFileInspector {
+	return gcsInspector{
+		qemuClient: NewInfoClient(),
+		fuseClient: gcsfuse.NewClient()}
+}
+
+// gcsInspector implements virtualDiskFileInspector using qemu-img gcsfuse.
+type gcsInspector struct {
+	qemuClient InfoClient
+	fuseClient gcsfuse.Client
+}
+
+func (inspector gcsInspector) Inspect(ctx context.Context, gcsURI string) (metadata VirtualDiskFileMetadata, err error) {
+	operation := func() error {
+		metadata, err = inspector.inspectOnce(ctx, gcsURI)
+		return err
+	}
+	return metadata, backoff.Retry(operation,
+		backoff.WithContext(backoff.NewConstantBackOff(50*time.Millisecond), ctx))
+}
+
+func (inspector gcsInspector) inspectOnce(ctx context.Context, gcsURI string) (metadata VirtualDiskFileMetadata, err error) {
+	bucket, object, err := segmentGCSURI(gcsURI)
+	if err != nil {
+		return metadata, err
+	}
+	mountedDirectory, err := inspector.fuseClient.MountToTemp(ctx, bucket)
+	if err != nil {
+		return metadata, err
+	}
+	defer inspector.fuseClient.Unmount(mountedDirectory)
```

Comment:
Can mount actually succeed even if error is thrown? Consider running unmout even on attemted mount if so. We had a case where error doesn't mean the operation didn't succeed (like disk inflate). This might be a more clear cut case but something to think about.

Reply:
That's a good idea -- I don't see a downside to unconditionally unmounting.

Reply:
Updated in https://github.com/GoogleCloudPlatform/compute-image-tools/pull/1283/commits/2bcf36860c6d1a027db3c39fd65b5cf29e467eb0